st2common/st2common/util/crypto.py

```diff
@@ -13,40 +13,291 @@
+
+import os
+import json
 import binascii
+import base64
+
+from hashlib import sha1
+
+import six
+
+from cryptography.hazmat.primitives.ciphers import Cipher
+from cryptography.hazmat.primitives.ciphers import algorithms
+from cryptography.hazmat.primitives.ciphers import modes
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives import hmac
+from cryptography.hazmat.backends import default_backend
 
 __all__ = [
+    'KEYCZAR_HEADER_SIZE',
+    'KEYCZAR_AES_BLOCK_SIZE',
+    'KEYCZAR_HLEN',
+
     'read_crypto_key',
+
     'symmetric_encrypt',
-    'symmetric_decrypt'
+    'symmetric_decrypt',
+
+    'cryptography_symmetric_encrypt',
+    'cryptography_symmetric_decrypt',
+
+    # NOTE: Keyczar functions are here for testing reasons - they are only used by tests
+    'keyczar_symmetric_encrypt',
+    'keyczar_symmetric_decrypt',
+
+    'AESKey'
 ]
 
+# Keyczar related constants
+KEYCZAR_HEADER_SIZE = 5
+KEYCZAR_AES_BLOCK_SIZE = 16
+KEYCZAR_HLEN = sha1().digest_size
+
 
-def read_crypto_key(key_path, key_type=None):
+class AESKey(object):
+    """
+    Class representing AES key object.
     """
-    Return the crypto key given a path to key file and the key type.
 
-    :param key_path: Absolute path to file containing crypto key.
-    :type key_path: ``str``
+    aes_key_string = None
+    hmac_key_string = None
+    hmac_key_size = None
+    mode = None
+    size = None
 
-    :param key_type: Type of crypto key.
-    :type key_type: :class:`keyczar.keys.KeyType`
+    def __init__(self, aes_key_string, hmac_key_string, hmac_key_size, mode='CBC', size=128):
+        if mode not in 'CBC':
+            raise ValueError('Unsupported mode: %s' % (mode))
 
-    :rtype: ``str``
+        if size < 128:
+            raise ValueError('Unsafe key size: %s' % (size))
+
+        self.aes_key_string = aes_key_string
+        self.hmac_key_string = hmac_key_string
+        self.hmac_key_size = int(hmac_key_size)
+        self.mode = mode.upper()
+        self.size = int(size)
+
+        # We also store bytes version of the key since bytes are needed by encrypt and decrypt
+        # methods
+        self.hmac_key_bytes = Base64WSDecode(self.hmac_key_string)
+        self.aes_key_bytes = Base64WSDecode(self.aes_key_string)
+
+    @classmethod
+    def generate(self, key_size=256):
+        """
+        Generate a new AES key with the corresponding HMAC key.
+
+        :rtype: :class:`AESKey`
+        """
+        if key_size < 128:
+            raise ValueError('Unsafe key size: %s' % (key_size))
+
+        aes_key_bytes = os.urandom(int(key_size / 8))
+        aes_key_string = Base64WSEncode(aes_key_bytes)
+
+        hmac_key_bytes = os.urandom(int(key_size / 8))
+        hmac_key_string = Base64WSEncode(hmac_key_bytes)
+
+        return AESKey(aes_key_string=aes_key_string, hmac_key_string=hmac_key_string,
+                      hmac_key_size=key_size, mode='CBC', size=key_size)
+
+    def __json__(self):
```

Comment:
This is probably not a good idea at this point, since [special methods starting and ending with double underscores are reserved for the Python interpreter](https://bugs.python.org/issue27362#msg269002) (read: their names get mangled by CPython).

Is this for backwards compatibility with keyczar? That's really the only good reason I can think to use a dunder for this method.

I would rename this to something like `to_json` or `_to_json` if it needs to be private.

Reply:
Yeah, I agree - I will change it to `to_json` (`__json__` is kinda a left over from the simplejson library,).